### PR TITLE
add labels update endpoint and label colors

### LINKED ISSUE #1955: labels in a color

### PR DESCRIPTION
**Pull request by jademcgough**
Closes #1955 #1956 

TODO:
- [x] add endpoints in swagger
- [ ] store label colors in separate bucket so that it can be shared between tags with same name and orgID

### CODE REVIEW

File: label.go [hunk 1]
```diff
@@ -37,9 +48,19 @@ func (l *Label) Validate() error {
 		return errors.New("label name is required")
 	}
 
+	if l.Color != "" && !colorPattern.MatchString(l.Color) {
+		return errors.New("label color must be valid hex string")
+	}
+
 	return nil
```
Comment (kelwang) [inline label.go]: maybe use platform.Error here as well, Code: EInvalid

File: bolt/label.go [hunk 2]
```diff
@@ -131,10 +153,82 @@ func (c *Client) uniqueLabel(ctx context.Context, tx *bolt.Tx, l *platform.Label
...
+
+func (c *Client) updateLabel(ctx context.Context, tx *bolt.Tx, l *platform.Label, upd platform.LabelUpdate) (*platform.Label, error) {
+	ls, err := c.findLabels(ctx, tx, platform.LabelFilter{Name: l.Name, ResourceID: l.ResourceID})
+	if err != nil {
+		return nil, err
+	}
+	if len(ls) == 0 {
+		return nil, &platform.Error{
+			Code: platform.ENotFound,
+			Op:   getOp(platform.OpUpdateLabel),
```
Comment (kelwang) [inline bolt/label.go]: I would prefer to make this op at the exported UpdateLabel method, for this one, we can call it "bolt/updateLabel"

File: bolt/label.go [hunk 3]
```diff
@@ -131,10 +153,82 @@ func (c *Client) uniqueLabel(ctx context.Context, tx *bolt.Tx, l *platform.Label
...
+	label := ls[0]
+
+	if upd.Color != nil {
+		label.Color = *upd.Color
+	}
+
+	if err := label.Validate(); err != nil {
+		return nil, &platform.Error{
+			Code: platform.EInvalid,
+			Op:   OpPrefix + platform.OpUpdateLabel,
```
Comment (kelwang) [inline bolt/label.go]: same here

File: bolt/label.go [hunk 4]
```diff
@@ -131,10 +153,82 @@ func (c *Client) uniqueLabel(ctx context.Context, tx *bolt.Tx, l *platform.Label
...
+			Msg:  err.Error(),
```
Comment (kelwang) [inline bolt/label.go]: Let's use Err:err, to show a stack of errors.

File: http/label_service.go [hunk 5]
```diff
@@ -190,6 +175,63 @@ func decodePostLabelRequest(ctx context.Context, r *http.Request) (*postLabelReq
...
+type patchLabelRequest struct {
+	label *plat.Label
+	upd   plat.LabelUpdate
+}
+
+func decodePatchLabelRequest(ctx context.Context, r *http.Request) (*patchLabelRequest, error) {
+	params := httprouter.ParamsFromContext(ctx)
+	id := params.ByName("id")
+	if id == "" {
+		return nil, kerrors.InvalidDataf("url missing resource id")
```
Comment (kelwang) [inline http/label_service.go]: change this to &platform.Error{Code:platform.EInvalid, Msg:"url missing resource id"}

File: http/label_service.go [hunk 6]
```diff
@@ -190,6 +175,63 @@ func decodePostLabelRequest(ctx context.Context, r *http.Request) (*postLabelReq
...
+	}
+
+	name := params.ByName("name")
+	if name == "" {
+		return nil, kerrors.InvalidDataf("label name is missing")
```
Comment (kelwang) [inline http/label_service.go]: change this to &platform.Error{Code:platform.EInvalid, Msg:"label name is missing"}

File: bolt/label.go (hunk 2 above)
Comment (jademcgough) [reply bolt/label.go]: 👍 yeah, I like that too

**Pull request merged**